File: backend/internal/models/product.go
package models

import (
	"database/sql/driver"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/lib/pq"
)

// ResilientStringArray is a slice of strings that is extremely robust when scanning array values from PostgreSQL / Supabase
type ResilientStringArray []string

// Scan implements the sql.Scanner interface
func (a *ResilientStringArray) Scan(src interface{}) error {
	if src == nil {
		*a = make(ResilientStringArray, 0)
		return nil
	}

	switch v := src.(type) {
	case []byte:
		return a.scanBytes(v)
	case string:
		return a.scanBytes([]byte(v))
	default:
		// Fallback to standard pq.StringArray scanner
		var psa pq.StringArray
		if err := psa.Scan(src); err == nil {
			*a = ResilientStringArray(psa)
			return nil
		}
		return fmt.Errorf("unsupported source type for ResilientStringArray: %T", src)
	}
}

func (a *ResilientStringArray) scanBytes(b []byte) error {
	if len(b) == 0 {
		*a = make(ResilientStringArray, 0)
		return nil
	}

	str := string(b)

	// Format 1: JSON array e.g. ["img1", "img2"]
	if strings.HasPrefix(str, "[") && strings.HasSuffix(str, "]") {
		var slice []string
		if err := json.Unmarshal(b, &slice); err == nil {
			*a = ResilientStringArray(slice)
			return nil
		}
	}

	// Format 2: PostgreSQL native array e.g. {img1,img2}
	if strings.HasPrefix(str, "{") && strings.HasSuffix(str, "}") {
		var psa pq.StringArray
		if err := psa.Scan(b); err == nil {
			*a = ResilientStringArray(psa)
			return nil
		}
	}

	// Format 3: Comma separated values
	if str != "" {
		parts := strings.Split(str, ",")
		for i := range parts {
			parts[i] = strings.Trim(parts[i], `" `)
		}
		*a = ResilientStringArray(parts)
		return nil
	}

	*a = make(ResilientStringArray, 0)
	return nil
}

// Value implements the driver.Valuer interface
func (a ResilientStringArray) Value() (driver.Value, error) {
	if len(a) == 0 {
		return "{}", nil
	}
	return pq.StringArray(a).Value()
}

// Product model - inherits BaseEntity through embedding
type Product struct {
	BaseEntity
	CompanyID        string               `json:"company_id" db:"company_id"`
	Name             string               `json:"name" db:"name"`
	Description      string               `json:"description" db:"description"`
	HSCode           *string              `json:"hs_code,omitempty" db:"hs_code"`
	HSCodeConfidence *float64             `json:"hs_code_confidence,omitempty" db:"hs_code_confidence"`
	PriceRangeMin    *float64             `json:"price_range_min,omitempty" db:"price_range_min"`
	PriceRangeMax    *float64             `json:"price_range_max,omitempty" db:"price_range_max"`
	Currency         string               `json:"currency" db:"currency"`
	MOQ              *int                 `json:"moq,omitempty" db:"moq"`
	Images           ResilientStringArray `json:"images" db:"images"`
	Category         string               `json:"category" db:"category"`
	CountryOrigin    string               `json:"country_origin" db:"country_origin"`
	ListingScore     *int                 `json:"listing_score,omitempty" db:"listing_score"`
	ViewCount        int                  `json:"view_count" db:"view_count"`
	InquiryCount     int                  `json:"inquiry_count" db:"inquiry_count"`
}

// SetPriceRange sets price range with validation
func (p *Product) SetPriceRange(min, max float64, currency string) error {
	if min > max {
		return ErrInvalidPriceRange
	}
	p.PriceRangeMin = &min
	p.PriceRangeMax = &max
	p.Currency = currency
	return nil
}

// IncrementViewCount increments view counter
func (p *Product) IncrementViewCount() {
	p.ViewCount++
}

// IncrementInquiryCount increments inquiry counter
func (p *Product) IncrementInquiryCount() {
	p.InquiryCount++
}

// CalculateListingScore calculates product listing completeness score
func (p *Product) CalculateListingScore() int {
	score := 0

	if len(p.Name) > 10 {
		score += 20
	}
	if len(p.Description) > 50 {
		score += 20
	}
	if p.HSCode != nil {
		score += 20
	}
	if len(p.Images) >= 3 {
		score += 20
	}
	if p.PriceRangeMin != nil && p.PriceRangeMax != nil {
		score += 20
	}

	p.ListingScore = &score
	return score
}

File: frontend/src/models/Product.ts
import { BaseEntityModel } from './BaseEntity';
import { Product as IProduct } from '@/types/product';

/**
 * Product Model
 * Implements Encapsulation & Inheritance (extends BaseEntityModel)
 */
export class ProductModel extends BaseEntityModel {
  private _company_id: string;
  private _name: string;
  private _description: string;
  private _hs_code?: string;
  private _hs_code_confidence?: number;
  private _price_range_min?: number;
  private _price_range_max?: number;
  private _currency: string;
  private _moq?: number;
  private _images: string[];
  private _category: string;
  private _country_origin: string;
  private _listing_score?: number;
  private _view_count: number;
  private _inquiry_count: number;

  constructor(data: IProduct) {
    super(data);
    this._company_id = data.company_id;
    this._name = data.name;
    this._description = data.description;
    this._hs_code = data.hs_code;
    this._hs_code_confidence = data.hs_code_confidence;
    this._price_range_min = data.price_range_min;
    this._price_range_max = data.price_range_max;
    this._currency = data.currency || 'USD';
    this._moq = data.moq;
    this._images = data.images || [];
    this._category = data.category;
    this._country_origin = data.country_origin;
    this._listing_score = data.listing_score;
    this._view_count = data.view_count || 0;
    this._inquiry_count = data.inquiry_count || 0;
  }

  // Getters (Encapsulation)
  get companyId(): string {
    return this._company_id;
  }

  get name(): string {
    return this._name;
  }

  get description(): string {
    return this._description;
  }

  get hsCode(): string | undefined {
    return this._hs_code;
  }

  get priceRange(): string {
    if (!this._price_range_min || !this._price_range_max) {
      return 'Contact for pricing';
    }
    return `${this._currency} ${this._price_range_min.toLocaleString()} - ${this._price_range_max.toLocaleString()}`;
  }

  get listingScore(): number {
    return this._listing_score || 0;
  }

  get images(): string[] {
    return this._images;
  }

  get category(): string {
    return this._category;
  }

  // Setters (Encapsulation with validation)
  setHSCode(code: string, confidence: number): void {
    this._hs_code = code;
    this._hs_code_confidence = confidence;
  }

  setPriceRange(min: number, max: number, currency: string = 'USD'): void {
    if (min > max) {
      throw new Error('Minimum price cannot be greater than maximum price');
    }
    this._price_range_min = min;
    this._price_range_max = max;
    this._currency = currency;
  }

  incrementViewCount(): void {
    this._view_count++;
  }

  incrementInquiryCount(): void {
    this._inquiry_count++;
  }

  // Business logic methods
  isListingComplete(): boolean {
    return !!(
      this._name &&
      this._description &&
      this._hs_code &&
      this._images.length > 0 &&
      this._price_range_min &&
      this._price_range_max
    );
  }

  calculateListingScore(): number {
    let score = 0;
    
    if (this._name && this._name.length > 10) score += 20;
    if (this._description && this._description.length > 50) score += 20;
    if (this._hs_code) score += 20;
    if (this._images.length >= 3) score += 20;
    if (this._price_range_min && this._price_range_max) score += 20;

    this._listing_score = score;
    return score;
  }

  toJSON(): IProduct {
    return {
      id: this._id,
      created_at: this._created_at,
      updated_at: this._updated_at,
      deleted_at: this._deleted_at,
      company_id: this._company_id,
      name: this._name,
      description: this._description,
      hs_code: this._hs_code,
      hs_code_confidence: this._hs_code_confidence,
      price_range_min: this._price_range_min,
      price_range_max: this._price_range_max,
      currency: this._currency,
      moq: this._moq,
      images: this._images,
      category: this._category,
      country_origin: this._country_origin,
      listing_score: this._listing_score,
      view_count: this._view_count,
      inquiry_count: this._inquiry_count,
    };
  }
}
